File: LeetCode/p1115_print_foobar_semaphore.py
from threading import Semaphore

class FooBar:
    def __init__(self, n):
        self.n = n
        self.can_foo = Semaphore(1)
        self.can_bar = Semaphore(0)


    def foo(self, printFoo: 'Callable[[], None]') -> None:
        
        for i in range(self.n):
            self.can_foo.acquire()
            # printFoo() outputs "foo". Do not change or remove this line.
            printFoo()
            self.can_bar.release()

    def bar(self, printBar: 'Callable[[], None]') -> None:
        
        for i in range(self.n):
            self.can_bar.acquire()
            # printBar() outputs "bar". Do not change or remove this line.
            printBar()
            self.can_foo.release()

File: LeetCode/p1115_print_foobar_semaphore.java
import java.util.concurrent.Semaphore;

class FooBar {
    private int n;

    Semaphore canFoo, canBar;

    public FooBar(int n) {
        this.n = n;
        this.canFoo = new Semaphore(1);
        this.canBar = new Semaphore(0);
    }

    public void foo(Runnable printFoo) throws InterruptedException {
        for (int i = 0; i < n; i++) {
            canFoo.acquire();
        	// printFoo.run() outputs "foo". Do not change or remove this line.
        	printFoo.run();
            canBar.release();
        }
    }

    public void bar(Runnable printBar) throws InterruptedException {
        for (int i = 0; i < n; i++) {
            canBar.acquire();
            // printBar.run() outputs "bar". Do not change or remove this line.
        	printBar.run();
            canFoo.release();
        }
    }
}
